### PR TITLE
feature: finish upgrade interface

### PR DESCRIPTION
**Pull request by HusterWan**
Signed-off-by: Michael Wan <zirenwan@gmail.com>

### Ⅰ. Describe what this PR did
finish upgrade interface daemon part
```
root@osboxes:wanziren -> pouch upgrade --name test1 registry.hub.docker.com/library/busybox:latest
test1
```

### Ⅳ. Describe how to verify it
* 1. Create a new container 
```
root@osboxes:pouch (zr/finish-upgrade) -> pouch inspect test2
{
    "Args": null,
    "Config": {
        "Cmd": [
            "sh"
        ],
        "Entrypoint": null,
        "Env": [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        ],
        "Image": "registry.hub.docker.com/library/busybox:latest",
        "OnBuild": null,
        "Shell": null
    },
    "Created": "2018-03-20T02:19:05.418084614Z",
    "HostConfig": {
        "NetworkMode": "bridge",
        "RestartPolicy": {
            "Name": "no"
        },
        "Runtime": "runc",
        "BlkioDeviceReadBps": null,
        "BlkioDeviceReadIOps": null,
        "BlkioDeviceWriteBps": null,
        "BlkioDeviceWriteIOps": null,
        "BlkioWeightDevice": null,
        "CgroupParent": "default",
        "DeviceCgroupRules": null,
        "Devices": [],
        "Memory": 20971520,
        "MemoryExtra": 0,
        "MemorySwappiness": -1,
        "MemoryWmarkRatio": 0,
        "Ulimits": null
    },
    "Id": "4c58d27f58d38776dda31c01c897bbf554c802a9b80ae4dc20be1337f8a969f2",
    "Image": "registry.hub.docker.com/library/busybox:latest",
    "Mounts": null,
    "Name": "test2",
    "NetworkSettings": {
        "Networks": {
            "bridge": {
                "Aliases": null,
                "EndpointID": "e3845861204350e6ec1ac2fbfaece7d3fe384ed0f5c01b41b53478d89b70a131",
                "Gateway": "172.17.0.1",
                "IPAddress": "172.17.0.2/24",
                "IPPrefixLen": 24,
                "Links": null,
                "MacAddress": "02:42:ac:11:00:02",
                "NetworkID": "59a8a8a365171ea1862f11534be913a81a6b4fc39f516b3121fb71c52b3dc2f8"
            }
        },
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null
    },
    "State": {
        "Pid": 4895,
        "StartedAt": "2018-03-20T02:19:05.780853724Z",
        "Status": "running"
    }
}
```

* 2. Update container's image
```
root@osboxes:pouch (zr/finish-upgrade) -> pouch upgrade --name test2 registry.hub.docker.com/library/ubuntu:14.04
test2
```

* 3. Check the container's status
```
root@osboxes:pouch (zr/finish-upgrade) -> pouch ps
Name    ID       Status          Created          Image                                                Runtime
test2   4c58d2   Up 2 minutes    15 minutes ago   registry.hub.docker.com/library/ubuntu:14.04         runc
```

* 4. Inspect the container's meta info
```
root@osboxes:pouch (zr/finish-upgrade) -> pouch inspect test2
{
    "Args": null,
    "Config": {
        "Cmd": [
            "sh"
        ],
        "Entrypoint": null,
        "Env": [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        ],
        "Image": "registry.hub.docker.com/library/ubuntu:14.04",
        "OnBuild": null,
        "Shell": null
    },
    "Created": "2018-03-20T02:19:05.418084614Z",
    "HostConfig": {
        "NetworkMode": "bridge",
        "RestartPolicy": {
            "Name": "no"
        },
        "Runtime": "runc",
        "BlkioDeviceReadBps": null,
        "BlkioDeviceReadIOps": null,
        "BlkioDeviceWriteBps": null,
        "BlkioDeviceWriteIOps": null,
        "BlkioWeightDevice": null,
        "CgroupParent": "default",
        "DeviceCgroupRules": null,
        "Devices": [],
        "Memory": 20971520,
        "MemoryExtra": 0,
        "MemorySwappiness": -1,
        "MemoryWmarkRatio": 0,
        "Ulimits": null
    },
    "Id": "4c58d27f58d38776dda31c01c897bbf554c802a9b80ae4dc20be1337f8a969f2",
    "Image": "registry.hub.docker.com/library/ubuntu:14.04",
    "Mounts": null,
    "Name": "test2",
    "NetworkSettings": {
        "Networks": {
            "bridge": {
                "Aliases": null,
                "EndpointID": "e3845861204350e6ec1ac2fbfaece7d3fe384ed0f5c01b41b53478d89b70a131",
                "Gateway": "172.17.0.1",
                "IPAddress": "172.17.0.2/24",
                "IPPrefixLen": 24,
                "Links": null,
                "MacAddress": "02:42:ac:11:00:02",
                "NetworkID": "59a8a8a365171ea1862f11534be913a81a6b4fc39f516b3121fb71c52b3dc2f8"
            }
        },
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null
    },
    "State": {
        "Pid": 8251,
        "StartedAt": "2018-03-20T02:32:43.874327747Z",
        "Status": "running"
    }
}
```

### CODE REVIEW

**Comment by allencloud** [comment]
Thanks a lot for the work you made. I am wondering adding some integration test will guarantee this  much better. @HusterWan 💟

File: daemon/mgr/container.go [hunk 1]
```diff
@@ -855,7 +863,126 @@ func (mgr *ContainerManager) Update(ctx context.Context, name string, config *ty
...
+		return err
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	// check the image existed or not, and convert image id to image ref
+	image, err := mgr.ImageMgr.GetImage(ctx, config.Image)
+	if err != nil {
+		return errors.Wrap(err, "get image failed")
```
Comment (allencloud) [inline daemon/mgr/container.go]: failed to get image

File: daemon/mgr/container.go [hunk 2]
```diff
@@ -855,7 +863,126 @@ func (mgr *ContainerManager) Update(ctx context.Context, name string, config *ty
...
+	}
+
+	// FIXME: image.Name does not exist,so convert Repotags or RepoDigests to ref
+	ref := ""
+	if len(image.RepoTags) > 0 {
+		ref = image.RepoTags[0]
+	} else {
+		ref = image.RepoDigests[0]
```
Comment (allencloud) [inline daemon/mgr/container.go]: Is there possibility that this `image.RepoDigests` has a length of 0? 
also ping @YaoZengzeng @zeppp

File: daemon/mgr/container.go [hunk 3]
```diff
@@ -855,7 +863,126 @@ func (mgr *ContainerManager) Update(ctx context.Context, name string, config *ty
...
+	var (
+		needRollback        = false
+		backupContainerMeta = *c.meta
+	)
+
+	defer func() {
+		if needRollback {
+			c.meta = &backupContainerMeta
+			if err := mgr.Client.CreateSnapshot(ctx, c.ID(), c.meta.Image); err != nil {
+				logrus.Errorf("Upgrade rollback failed to create snapshot: %v", err)
```
Comment (allencloud) [inline daemon/mgr/container.go]: failed to create snapshot when rollback upgrade: %v

File: daemon/mgr/container.go [hunk 4]
```diff
@@ -855,7 +863,126 @@ func (mgr *ContainerManager) Update(ctx context.Context, name string, config *ty
...
+		// wait util old snapshot to be deleted
+		wait := make(chan struct{})
+		go func() {
+			for {
+				// FIXME(ziren) Ensure the removed snapshot be removed
+				// by garbage collection.
+				time.Sleep(100 * time.Millisecond)
+
+				msg, err := mgr.Client.GetSnapshot(ctx, c.ID())
+				logrus.Info("snapshots %+v", msg)
```
Comment (allencloud) [inline daemon/mgr/container.go]: If err != nil, just double check whether msg has data in itself.

File: test/cli_upgrade_test.go [hunk 5]
```diff
@@ -29,3 +35,115 @@ func (suite *PouchUpgradeSuite) SetUpSuite(c *check.C) {
...
 }
+
+// TestPouchUpgrade is to verify pouch upgrade command.
+func (suite *PouchUpgradeSuite) TestPouchUpgrade(c *check.C) {
+	name := "TestPouchUpgradeSuite"
+
+	command.PouchRun("run", "-d", "--name", name, busyboxImage).Assert(c, icmd.Success)
+
+	res := command.PouchRun("upgrade", "--name", name, busyboxImage)
+	c.Assert(res.Error, check.IsNil)
```
Comment (allencloud) [inline test/cli_upgrade_test.go]: I am wondering if we do not upgrade anything, should we return an error indicating `nothing to upgrade` and no status update happens in daemon side. @HusterWan WDYT

File: daemon/mgr/container.go (hunk 4 above)
Comment (HusterWan) [reply daemon/mgr/container.go]: if error occured , msg will be a empty Snapshot.Info{}

**Comment by allencloud** [comment]
CI fails:
```
FAIL: /go/src/github.com/alibaba/pouch/test/cli_upgrade_test.go:58: PouchUpgradeSuite.TestPouchUpgradeNoChange
/go/src/github.com/alibaba/pouch/test/cli_upgrade_test.go:64:
    c.Assert(res.Error, check.NotNil)
... value = nil
```
@HusterWan

**Comment by allencloud** [comment]
LGTM
And I think we need to add a document to tell users how to take advantages of this API. WDYT? @HusterWan

**Comment by HusterWan** [comment]
> And I think we need to add a document to tell users how to take advantages of this API. WDYT? @HusterWan

Yes, i will add the doc after i have a tiime, but i will create a issue to reminder me.

**Pull request merged**